File: backend/bin/main/application/services/UserSettingsService.kt
/*
 * ॐ नमः शिवाय
 * By: Sheersh Bhatnagar
 * Date: 27/05/26 21:07
 */

package dev.sheershbhatnagar.ai_assistant.application.services

import java.time.LocalDateTime
import java.util.UUID

import dev.sheershbhatnagar.ai_assistant.domain.models.UserSettings
import dev.sheershbhatnagar.ai_assistant.domain.repository.AiModelRepository
import dev.sheershbhatnagar.ai_assistant.domain.repository.UserSettingsRepository

class UserSettingsService(
    private val userSettingsRepository: UserSettingsRepository,
    private val aiModelRepository: AiModelRepository
) {
    suspend fun getUserSettings(userId: UUID): UserSettings? {
        return userSettingsRepository.getSettingsByUserId(userId)
    }

    suspend fun updateDefaultModel(userId: UUID, defaultModelId: UUID): UserSettings {
        // Security check: Verify the model configuration exists and belongs to the requesting user
        val model = aiModelRepository.getModelById(defaultModelId)
            ?: throw IllegalArgumentException("Model configuration not found")

        if (model.userId != userId) {
            throw IllegalArgumentException("Model configuration does not belong to the user")
        }

        val existingSettings = userSettingsRepository.getSettingsByUserId(userId)
        val now = LocalDateTime.now()

        val settingsToSave = if (existingSettings != null) {
            existingSettings.copy(
                defaultModelId = defaultModelId,
                updatedAt = now
            )
        } else {
            UserSettings(
                id = UUID.randomUUID(),
                userId = userId,
                defaultModelId = defaultModelId,
                createdAt = now,
                updatedAt = now
            )
        }

        return userSettingsRepository.upsertSettings(settingsToSave)
            ?: throw IllegalStateException("Failed to update user settings")
    }
}

File: backend/app/services/user_settings_service.py
from datetime import datetime
from typing import Optional
from sqlalchemy.orm import Session
from app.models.user_settings import UserSettings
from app.models.ai_model import AiModel

class UserSettingsService:
    def get_user_settings(self, db: Session, user_id: str) -> Optional[UserSettings]:
        return db.query(UserSettings).filter(UserSettings.user_id == user_id).first()

    def update_default_model(self, db: Session, user_id: str, default_model_id: str) -> UserSettings:
        # Check model exists
        model = db.query(AiModel).filter(AiModel.id == default_model_id).first()
        if not model:
            raise ValueError("Model configuration not found")

        # Security check: Model belongs to the user
        if str(model.user_id) != user_id:
            raise ValueError("Model configuration does not belong to the user")

        settings = db.query(UserSettings).filter(UserSettings.user_id == user_id).first()
        now = datetime.utcnow()

        if settings:
            settings.default_model_id = default_model_id
            settings._updated_at = now
        else:
            settings = UserSettings(
                user_id=user_id,
                default_model_id=default_model_id,
                _created_at=now,
                _updated_at=now
            )
            db.add(settings)

        db.commit()
        db.refresh(settings)
        return settings
